### DISCUSSION TITLE
Shell Method Order

### DISCUSSION THREAD

**Discussion by danvega** [spring-projects/spring-shell#636] [Q&A]
Is there a way to order the shell methods when you type help? It doesn't go by name and there doesn't seem to be any rhyme or reason to the ordering. When I type help I want to give the user something like 

1. Menu Option 1
2. Menu Option 2
3. Menu Option 3

**Comment by jvalkeal** [comment]
Currently no and you're not a first one asking about as just yday @markpollack asked a same question.

This might be a good moment to think about what users want as only thing you can customise is to hide group names(which essentially used for natural sorting for groups). Though it's possible to change antlr's stringtemplate files(which are used for rendering) in build in help command but that's a bit hardcore solution. Replacing existing help command is relatively easy but don't know how many would want to go through that exercise.

- Natural ordering to commands as well(what about reverse ordering?)
- Should there be a custom ordering possible based on some order number?
- Hide something from help(currently only hidden commands are skipped)?

**Comment by danvega** [comment]
Thank you for clarifying, just wasn't sure if I was missing something. I think what I might do is just print something to the console when the application starts that includes some ascii art + help text and then I can control the output.

**Comment by BillyBadBoy** [comment]
We're using a customized version of the `GroupsInfoModel` class in which we've replaced:
`Collectors.toMap(Entry::getKey, Entry::getValue)`, with
`Collectors.toMap(Entry::getKey, Entry::getValue, (v1, v2) -> v1, TreeMap::new)`

This gives use the commands listed in alphabetical order (within each group) when "help" is selected.

**Comment by zhoumiv** [comment]
You can achieve custom functionality for the `help` command by defining your own `Help` command:

```java
@ShellComponent
public class MyHelpCommand extends AbstractShellComponent implements Help.Command {

     // Just copy the implementation code from `org.springframework.shell.standard.commands.Help` here
}
```

Copy the following classes into the same package as `MyHelpCommand`:  
`GroupsInfoModel`, `GroupCommandInfoModel`, `CommandInfoModel`, `CommandParameterInfoModel`, and `CommandAvailabilityInfoModel`.  
These classes are all located in `org.springframework.shell.standard.commands`.  If you're using IntelliJ IDEA, multi-select these classes with your mouse, press "Cmd+C", then click on the package containing `MyHelpCommand` and press "Cmd+V" — the copy is complete.
After launching your Spring Shell application, you’ll find that the **`help`** command and now invokes `MyHelpCommand`.

1. To customize the order of command groups, modify the `of` method in `GroupsInfoModel`.  
   For example, to place `"Built-In Commands"` last:

```java
// collect commands into groups with sorting
...
SortedMap<String, Map<String, CommandRegistration>> commandsByGroupAndName = ...

// reorder groups so that 'Built-In' is last
Map<String, CommandRegistration> builtInCommands = commandsByGroupAndName.remove("Built-In Commands");
SequencedMap<String, Map<String, CommandRegistration>> groupToCommands = new LinkedHashMap<>(commandsByGroupAndName);
groupToCommands.putLast("Built-In Commands", builtInCommands);
```

2. To customize the order of commands *within* each group, continue modifying the `of` method in `GroupsInfoModel`.  
   Specifically, adjust how the `cims` list is built — for example, sort commands alphabetically by name:

```java
// build model
List<GroupCommandInfoModel> gcims =
        groupToCommands.entrySet()
                       .stream()
                       .map(e -> {
                           List<CommandInfoModel> cims =
                                   e.getValue()
                                    .entrySet().stream()
                                    .map(ee -> CommandInfoModel.of(ee.getKey(), ee.getValue()))
                                    .sorted()  // Let the `CommandInfoModel` implement `Comparable` (see below)
                                    .collect(Collectors.toList());
                           return GroupCommandInfoModel.of(e.getKey(), cims);
                       })
                       .collect(Collectors.toList());
```

```java
public class CommandInfoModel implements Comparable<CommandInfoModel> {
    ...

    @Override
    public int compareTo(CommandInfoModel that) {
        return this.name.compareTo(that.name);
    }
}
```

My Java version: 25.0.1
My Spring Shell version: 3.4.1

**Comment by fmbenhassine** [comment]
> This gives use the commands listed in alphabetical order (within each group) when "help" is selected.

This is what I suggest to implement as a built-in feature in Spring Shell in https://github.com/spring-projects/spring-shell/issues/1279#issuecomment-3785926026, no need to modify GroupsInfoModel or [copy/paste classes or defining your own Help command](https://github.com/spring-projects/spring-shell/discussions/636#discussioncomment-15606483).

I will close this discussion in favor of #1279
